doc/design/block.md

```diff
@@ -0,0 +1,210 @@
+
+To store the definition of Variables and Operators, a C++ class `SymbolTable` is introduced as a similar concept with compiler's symbol table.
+
+`SymbolTable` will has following functions:
+
+- store the definitions (some names and attributes) of variables and operators,
+- to verify if a variable name has been declared,
+- to make it possible to implement type checking (offer Protobuf message pointers to `InferShape` handlers).
+
+
+```c++
+// Information in SymbolTable is enough to trace the dependency graph. So maybe
+// the Eval() interface takes a SymbolTable is enough.
+class SymbolTable {
+ public:
+  SymbolTable(SymbolTable* parent) : parent_(parent) {}
+
+  OpDesc* NewOp(const string& name="");
+
+  // TODO determine whether name is generated by python or C++
+  // currently assume that a unique name will be generated by C++ if the
+  // argument name left default.
+  VarDesc* NewVar(const string& name="");
+
+  // find a VarDesc by name, if recursive true, find parent's SymbolTable
+  // recursively.
+  // this interface is introduced to support InferShape, find protobuf messages
+  // of variables and operators, pass pointers into InferShape.
+  // operator
+  //
+  // NOTE maybe some C++ classes such as VarDescBuilder and OpDescBuilder should
+  // be proposed and embedded into pybind to enable python operate on C++ pointers.
+  VarDesc* FindVar(const string& name, bool recursive=true);
+
+  OpDesc* FindOp(const string& name);
+
+  BlockDesc Serialize() const;
+  void DeSerialize();
+
+ private:
+  SymbolTable* parent_;
+
+  map<string, OpDesc> ops_;
+  map<string, VarDesc> vars_;
+};
+```
+
+After all the description of variables and operators is added into SymbolTable,
+the block has enough information to run.
+
+To make block simpler, we use another class `RuntimeTable` to create Variable and Operator objects. `RuntimeTable` is free to add memory optimizements in the future.
+
+The `Block` class takes a `BlockDesc` as input, and provide `Run` and `InferShape` functions.
+
+
+```c++
+namespace {
+// A RuntimeTable maintains all the runtime environment symbols of a block,
+// including the variables and operators.
+// NOTE the implementation of runtime environment is seperated from Block
+// to make it easier to improve memory usage and some other improvements
+// based on a block's description.
+// NOTE this name is aweful, and should be updated latter.
+class RuntimeTable {
+ public:
+  // scope: the scope to create new variables.
+  RuntimeTable(const BlockDesc& desc, const Scope& scope);
+
+  vector<OperatorBase*> ops() { return ops_; }
+
+  bool IsValid() const { return ops_.empty(); }
+
+ protected:
+  void CreateVars(const Scope& scope);
+  void CreateOps();
+
+ private:
+  SymbolTable symbol_tabel_;
+  vector<OperatorBase*> ops_;
+};
+}
+
+
+class Block : OperatorBase {
+public:
+  // desc may be serialized from a RuntimeTable or not.
```

Comment:
Where is the method that creates a `BlockDesc` from `RuntimeTable`?

Reply:
RuntimeTable will be merged into Block as its Construction, it takes a `BlockDesc` as input, and creates operator instances for Block.